<!doctype html>  
<html lang="en">
	
	<head>
		<meta charset="utf-8">
		
		<title>reveal.js</title>

		<meta name="description" content="An easy to use CSS 3D slideshow tool for quickly creating good looking HTML presentations.">
		<meta name="author" content="Hakim El Hattab">

		<meta name="apple-mobile-web-app-capable" content="yes" />
		<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />
		
		<link href='http://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic' rel='stylesheet' type='text/css'>
		
		<link rel="stylesheet" href="css/reset.css">
		<link rel="stylesheet" href="css/main.css">
		<link rel="stylesheet" href="css/print.css" type="text/css" media="print">

		<link rel="stylesheet" href="lib/zenburn.css">
	</head>
	
	<body>
		
		<div class="reveal">

			<!-- Used to fade in a background when a specific slide state is reached -->
			<div class="state-background"></div>
			
			<!-- Any section element inside of this container is displayed as a slide -->
			<div class="slides">
                <section>
                    <h2>Aperture Laboratories Presents</h2> 
					<h2>Using Residue Conservation to Predict Protein Interactions</h2>
                    <h4 class="inverted">With our helpful science team:
                        </br>
                        James Bucher,</br>
                        Kyle Mears,</br>
                        Tyler Land,</br>
                        Christina Guiterrez</br>
                        and Ali Hijy</h4>
				</section>
				
                <section>
                    <section>
					<h2>Background</h2>
                    <p>
                    Motivation: </br >
                    <a href="http://www.sciencemag.org/content/286/5438/295.short">Evolutionarily conserved pathways of energetic connectivity in protein families</a> 
                    </br>
                    </br>
                    <a href="http://www.proteinscience.org/details/journalArticle/112256/Physicochemical_and_residue_conservation_calculations_to_improve_the_ranking_of_.html">Physicochemical and residue conservation calculations to improve the ranking of protein-protein docking solutions</a>
                    </p>
                    </section>
                    <section>
                        <h1>Their Idea</h1>
                        <h2>On a single protein, identify conserved regions that mutate together</h2>
                        <ul>    
                            <li>Look at many proteins from the same domain</li>
                            <li>Calculate a "change in free energy" for each pair of sites</li>
                            <li>Use this to measure covariance between proteins</li> 
                         </ul>
                    </section>
                    <section style="font-size:.75em">
                        <h1>Our Idea</h1>
                        <h2>For 2 Proteins</h2>
                        <ul>
                            <li>Identify conserved amino acids on each</li>
                            <li>Measure how one protein's conserved regions mutate with the other</li>
                            <li>Look at many proteins from multiple domains</li>
                            <li>Conserved regions that mutate together should also interact together</li>
                        </ul>
                    </section>
                </section>
                <section style="font-size:.850em">
                    <h1>The Process</h1>
                    <ol>
                        <li>Begin with a pair of proteins A and B</li>
                        <li>Search blast for proteins homologs to A and B</li>
                        <li>Align all of the homologs</li>
                        <li>For each mutation in every protein homologs to A, form the cross product with the mutations in proteins homologs to B</li>
                        <li>Only consider pairs of mutations that belong to the same species</li>
                        <li>Score each pair of mutations</li>
                        <li>Use those scores to score protein pairs that those mutations are found on</li>
                        <li>Normalize the score</li>
                    </ol>
                </section>
                <section>
                    <section>
                    <h1>Data</h1>
                    <ul>
                        <li>Start with two proteins</li>
                        <li>For each one:
                            <ul>
                                <li>Do a BLAST search to get homologues</li>
                                <li>Pre-Process them to fit our needs</li>
                                <li>Run a multi-alignment for the proper ones</li>
                            </ul>
                        </li>
                    </ul>
                    </section>
                    <section style="font-size:.85em">
                        <h1>Pre-Processing</h1>
                        <ul>
                            <li>Blast returns lots of proteins</li>
                            <li>Each protein returned has an E-Val
                                <ul>
                                    <li> The lower the E-Val the more "significant" the match is</li>
                                    <li> Filter Out E-Val > 1x10<sup>-5</sup></li>
                                </ul>
                            </li>
                            <li>Remove proteins that are:
                                <ul>
                                    <li>Predicted</li>
                                    <li>Pseudo-Gene</li>
                                    <li>Partial</li>
                                    <li>Synthetic</li>
                                </ul>
                            </li>
                            <li> Finally Make sure there is only one homologue per organism</li>
                        </ul>
                    </section>
                </section>
                <section>
                    <section>
                        <h1>Methods</h1>
                        <h2>How do we use these homologues to derive protein-protein interactions?</h2>
                    </section>
                    <section>
                        <img src="img/Methods-1.png" width="100%">
                    </section>
                    <section>
                        <img src="img/Methods-2.png" width="100%">
                    </section>
                    <section>
                        <img src="img/Methods-3.png" height="100%">
                    </section>
                    <section>
                        <img src="img/Methods-4.png" width="100%">
                    </section>
                    <section>
                        <img src="img/Methods-5.gif" width="100%">
                    </section>
                    <section>
                        <h1>More Math and Calculations!</h1>
                        <ol>
                            <li>Calculate a distance score: <sup>1</sup>/<sub>(1 + |Score_A + Score_B|)</sub></li>
                            <li>Filter out pairs with:
                                <ul>
                                    <li>Too low a score</li>
                                    <li>Scores that are in the middle and likely random</li>
                                </ul>
                            </li>
                            <li>Compensate for the size of the protein
                                <ul>
                                    <li>final_score = </br> <sup>(sum_of_scores x 100000)</sup>&frasl;<sub>(min(size_A,size_B)<sup>2</sup> x common_species_count)</sub></li>
                                </ul>
                            </li>
                        </ol>
                    </section>
                </section>
                <section>
                    <section>
                    <h1>Results</h1>
                    <table>
                        <tr>
                            <td>Interacting Scores:
                                <ul>
                                    <li>Mean = 19.8</li>
                                    <li>Median = 5.4</li>
                                    <li>Max = 76.7</li>
                                    <li>Min = 2.1</li>
                                </ul>
                            </td>
                            <td>Not Interacting Scores:
                                <ul>
                                    <li>Mean = 20.6</li>
                                    <li>Median = 5.3</li>
                                    <li>Max = 278.7</li>
                                    <li>Min = 0</li>
                                </ul>
                            </td>
                        </tr>
                    </table>
                    </section>
                        
                    <section>
                        <img src="img/Results-1.png" width="100%"> 
                    </section>
                    <section>
                        <img src="img/Results-2.png" width="100%">
                    </section>
                    <section>
                        <img src="img/Results-3.png" width="100%">
                    </section>
                    <section>
                        <img src="img/Results-4.png" width="100%">
                    </section>
                    <section>
                        <img src="img/Results-5.png" width="100%">
                    </section>
                </section>
                <section>
                    <h1>Accomplished Milestones</h1>
                    <ul>
                        <li>Biopython script that preforms BLAST searches</br>
                        using fasta files as input and pre-processes the data</li>
                        <li>Program that scores conserved regions:
                            <ul>
                                <li>Scales with Hadoop</li>
                                <li>Scoring algorithm itself is non-trivial</li>
                            </ul>
                        </li>
                    </ul>
                </section>
                <section>
                    <section>
                    <h1>Learning Experience</h1>
                    There are many variations on scoring 
                    </br>
                    </br>
                    <ul>
                        <li>Inter-protein analysis is harder than Intra-protein</li>
                        <li>Rigorous statistical methods very useful</li>
                        <li>Which is best? Lots of considerations</li>
                    </ul> 
                    </br>
                    </br>
                    Kyle learned to read some python! 
                    </br>
                    </br>
                    Most importantly: Inter-Protein Coevolution analysis is hard!
                    </section>
                    <section style="font-size:.85em">
                        <h1>Why is it so hard?</h1>
                        <a href="http://www.mendeley.com/research/a-novel-method-for-detecting-intramolecular-coevolution-adding-a-further-dimension-to-selective-constraints-analyses/">CAPS</a> paper states that many parameters should be considered.
                        </br>
                        </br>
                        <ul>
                            <li>Time since divergence</li>
                            <li>Biochemistry
                                <ul>
                                    <li>Solvent Accessibility: Is water interacting with the amino acid?</li>
                                    <li>Physiochemical amino acid properties</li>
                                </ul>
                            </li>
                            <li>Different proteins require different methods</li>
                            <li>Our scoring system has a lot of room for improvement</li>
                        </ul>
                    </section>    
                </section>
                <section  style="font-size:.75em">
                    <h1>Future Work</h1>
                    <h3>Currently our method detects conservation over the entire protein</h3>
                    <ul>
                        <li>A better method would detect each conserved region on the proteins</li>
                    </ul>
                    </br>
                    </br>
                    <h3>More Data is needed to improve accuracy</h3>
                    <ul>
                        <li>We want proteins that we're sure co-evolved</li>
                        <li>How does time since divergence affect our output?</li>
                        <li>How can time best be approximated?</li>
                    </ul>
                </section>
                <section>
                    <h3>Aperture Laboratories thanks you for your participation in this experiment </br>
                    </br>
                    We do what we must, because we can</h3>
                    </br>
                    <h4>Please remember, we are not responsible for any strange afflictions caused by your participation in today's event. 
                        </br>
                        </br>
                        If you have any questions please ask our friendly science team.</h4>
                </section>
			</div>

			<!-- The navigational controls UI -->
			<aside class="controls">
				<a class="left" href="#">&#x25C4;</a>
				<a class="right" href="#">&#x25BA;</a>
				<a class="up" href="#">&#x25B2;</a>
				<a class="down" href="#">&#x25BC;</a>
			</aside>

			<!-- Displays presentation progress, max value changes via JS to reflect # of slides -->
			<div class="progress"><span></span></div>
			
		</div>

		<!-- Optional libraries for code syntax highlighting and classList support in IE9 -->
		<script src="lib/highlight.js"></script>
		<script src="lib/classList.js"></script>
		
		<script src="js/reveal.js"></script>
		
		<script>
			// Parse the query string into a key/value object
			var query = {};
			location.search.replace( /[A-Z0-9]+?=(\w*)/gi, function(a) {
				query[ a.split( '=' ).shift() ] = a.split( '=' ).pop();
			} );

			// Fires when a slide with data-state=customevent is activated
			Reveal.addEventListener( 'customevent', function() {
				alert( '"customevent" has fired' );
			} );

			// Fires each time a new slide is activated
			Reveal.addEventListener( 'slidechanged', function( event ) {
				// event.indexh & event.indexv
			} );

			Reveal.initialize({
				// Display controls in the bottom right corner
				controls: true,

				// Display a presentation progress bar
				progress: true,

				// If true; each slide will be pushed to the browser history
				history: true,

				// Loops the presentation, defaults to false
				loop: false,

				// Flags if mouse wheel navigation should be enabled
				mouseWheel: true,

				// Apply a 3D roll to links on hover
				rollingLinks: false,

				// UI style
				theme: query.theme || 'default', // default/neon

				// Transition style
				transition: query.transition || 'linear' // default/cube/page/concave/linear(2d)
			});

			hljs.initHighlightingOnLoad();
		</script>

	</body>
</html>
